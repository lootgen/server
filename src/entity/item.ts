import { Field, ID, ObjectType } from 'type-graphql';
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  PrimaryGeneratedColumn,
} from 'typeorm';

import { IItem } from './interfaces/item_interface';

@ObjectType()
@Entity()
export class Item extends BaseEntity implements IItem {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id: string;

  @Field()
  @Column({ unique: true })
  name: string;

  @Field()
  @CreateDateColumn()
  created: Date;
}
