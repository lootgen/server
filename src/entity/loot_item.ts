import { Field, ID, ObjectType } from 'type-graphql';
import {
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  PrimaryGeneratedColumn,
} from 'typeorm';

import { ILootItem } from './interfaces/loot_item_interface';

@ObjectType()
@Entity()
export class LootItem extends BaseEntity implements ILootItem {
  @Field((_type) => ID)
  @PrimaryGeneratedColumn('uuid')
  readonly id: string;

  @Field()
  @CreateDateColumn()
  readonly created: Date;

  @Field()
  @Column({ unique: true })
  name: string;
}
