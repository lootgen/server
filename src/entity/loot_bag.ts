import { Field, Int, ObjectType } from 'type-graphql';
import {
  AfterLoad,
  BaseEntity,
  Column,
  CreateDateColumn,
  Entity,
  JoinTable,
  ManyToMany,
  PrimaryGeneratedColumn,
} from 'typeorm';

import { ILootBag } from './interfaces/loot_bag_interface';
import { ILootItem } from './interfaces/loot_item_interface';
import { LootItem } from './loot_item';

@ObjectType()
@Entity()
export class LootBag extends BaseEntity implements ILootBag {
  @Field(() => Int)
  @PrimaryGeneratedColumn()
  readonly id: number;

  @Field()
  @CreateDateColumn()
  readonly created: Date;

  @Field((_type) => [LootItem])
  @ManyToMany('LootItem', {
    eager: true,
    cascade: true,
  })
  @JoinTable()
  items: ILootItem[];

  @Column({ nullable: false })
  raw: string;

  @AfterLoad()
  sortItems(): void {
    const sortedItems = this.raw.split('\n');
    this.items.sort((left, right) => {
      return sortedItems.indexOf(left.name) - sortedItems.indexOf(right.name);
    });
  }
}
