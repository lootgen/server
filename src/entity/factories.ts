import { ILike } from 'typeorm';

import { LootBag } from './loot_bag';
import { LootItem } from './loot_item';

export const makeLootBag = async (items: string[]): Promise<LootBag> => {
  const lootItems = await items.reduce(async (previous, item) => {
    const acc = await previous;
    try {
      const result = await LootItem.findOneOrFail({
        where: { name: ILike(item) },
      });
      return [...acc, result];
    } catch (_error) {
      const result = new LootItem();
      result.name = item;
      await result.save();
      return [...acc, result];
    }
  }, Promise.resolve([]));
  const result = new LootBag();
  result.items = lootItems;
  result.raw = lootItems.map((item) => item.name).join('\n');
  return await result.save();
};
