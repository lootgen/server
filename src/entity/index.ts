import { Connection } from 'typeorm';
import { LootBag } from './loot_bag';
import { LootItem } from './loot_item';

export const registerAllEntities = (connection: Connection): void => {
  LootBag.useConnection(connection);
  LootItem.useConnection(connection);
};
