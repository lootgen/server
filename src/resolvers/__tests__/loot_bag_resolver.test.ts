import { gql } from 'apollo-server';

import { makeLootBag } from '../../entity/factories';

const FETCH_LOOT_BAG = gql`
  query lootBag($id: Int!) {
    lootBag(input: { id: $id }) {
      items {
        item {
          name
        }
        order
      }
    }
  }
`;

const CREATE_LOOT_BAG_MUTATION = gql`
  mutation createLootBag($items: [String!]!) {
    createLootBag(input: { items: $items }) {
      id
      items {
        item {
          name
        }
        order
      }
    }
  }
`;

const ITEMS = ['1', '2', '3', '4', '5', '6', '7', '8'];

describe('Loot Bag resolver', () => {
  it('can fetch a loot bag by ID', async () => {
    const lootBag = await makeLootBag(ITEMS);
    const result = await global.client.query({
      query: FETCH_LOOT_BAG,
      variables: { id: lootBag.id },
    });
    expect(result).toMatchSnapshot();
  });

  it('sorts loot items based on insert order', async () => {
    const reversedItems = [...ITEMS].reverse();
    const lootBag = await makeLootBag(reversedItems);
    const result = await global.client.query({
      query: FETCH_LOOT_BAG,
      variables: { id: lootBag.id },
    });
    expect(result).toMatchSnapshot();
  });

  it('throws error: items must contain 8 elements', async () => {
    let result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items: [] },
    });
    expect(result).toMatchSnapshot();

    result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items: ITEMS.slice(0, 7) },
    });
    expect(result).toMatchSnapshot();

    result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items: [...ITEMS, '9'] },
    });
    expect(result).toMatchSnapshot();
  });

  it('throws error: loot items cannot be empty strings', async () => {
    const items = [...ITEMS];
    items.splice(0, 1, '');
    const result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items },
    });
    expect(result).toMatchSnapshot();
  });

  it('can create a new loot bag', async () => {
    const result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items: ITEMS },
    });
    expect(result).toMatchSnapshot();
  });

  it('can create a loot bag with duplicate items', async () => {
    const result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items: ['gm', 'gm', 'gm', 'gm', 'gm', 'gm', 'gm', 'gm'] },
    });
    expect(result).toMatchSnapshot();
  });

  it('throws error: loot items cannot contain profanity', async () => {
    const items = [...ITEMS];
    items.splice(4, 1, 'dumb fUcK');
    const result = await global.client.mutate({
      mutation: CREATE_LOOT_BAG_MUTATION,
      variables: { items },
    });
    expect(result).toMatchSnapshot();
  });
});
