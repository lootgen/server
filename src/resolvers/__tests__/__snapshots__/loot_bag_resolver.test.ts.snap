// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loot Bag resolver can create a new loot bag 1`] = `
Object {
  "data": Object {
    "createLootBag": Object {
      "id": 1,
      "items": Array [
        Object {
          "name": "1",
        },
        Object {
          "name": "2",
        },
        Object {
          "name": "3",
        },
        Object {
          "name": "4",
        },
        Object {
          "name": "5",
        },
        Object {
          "name": "6",
        },
        Object {
          "name": "7",
        },
        Object {
          "name": "8",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Loot Bag resolver can fetch a loot bag by ID 1`] = `
Object {
  "data": Object {
    "lootBag": Object {
      "items": Array [
        Object {
          "name": "1",
        },
        Object {
          "name": "2",
        },
        Object {
          "name": "3",
        },
        Object {
          "name": "4",
        },
        Object {
          "name": "5",
        },
        Object {
          "name": "6",
        },
        Object {
          "name": "7",
        },
        Object {
          "name": "8",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Loot Bag resolver sorts loot items based on insert order 1`] = `
Object {
  "data": Object {
    "lootBag": Object {
      "items": Array [
        Object {
          "name": "8",
        },
        Object {
          "name": "7",
        },
        Object {
          "name": "6",
        },
        Object {
          "name": "5",
        },
        Object {
          "name": "4",
        },
        Object {
          "name": "3",
        },
        Object {
          "name": "2",
        },
        Object {
          "name": "1",
        },
      ],
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Loot Bag resolver throws error: items must contain 8 elements 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Argument Validation Error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Loot Bag resolver throws error: items must contain 8 elements 2`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Argument Validation Error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Loot Bag resolver throws error: items must contain 8 elements 3`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Argument Validation Error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Loot Bag resolver throws error: loot items cannot be empty strings 1`] = `
Object {
  "data": null,
  "errors": Array [
    [GraphQLError: Argument Validation Error],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
